config/01-ce-router-defaults.conf

```
@@ -151,89 +150,83 @@ JOB_ROUTER_TRANSFORM_OrigRequests @=jrt
 JOB_ROUTER_TRANSFORM_Cpus @=jrt
     # Outside of the HTCondor WholeNodeJobs case, set RequestCpus to one of the following, in order:
     # 1. 'xcount' from the source job
-    # 2. RequestCpus from the source job
+    # 2. RequestCpus from the source job, but only if > 1
     # 3. default_xcount from the job route
     # 4. 1
+
+    if defined MY.xcount
+        EVALSET RequestCpus xcount
+    else
+        EVALMACRO test_RequestCpus ($(MY.RequestCpus:0)) > 1
+        if $(test_RequestCpus)
+            # RequestCpus already correct
+        else
+            EVALSET RequestCpus $(default_xcount:1)
+        endif
+    endif
+
+    # the old-style transforms ignored RequestCPUs for... reasons? when setting these. we do the same here.
+    EVALSET remote_SMPGranularity  xcount ?: ($(default_xcount:1))
+    COPY    remote_SMPGranularity  remote_NodeNumber
+
+    SET JobIsRunning           (JobStatus =!= 1) && (JobStatus =!= 5)
+
     if $(test_want_whole_node)
-        SET JOB_GLIDEIN_Cpus       "$$(MY.TotalCpus ?: JobCpus)"
+        SET JOB_GLIDEIN_Cpus       "$$([MY.TotalCpus ?: JobCpus])"
         # MATCH_EXP_JOB_GLIDEIN_Cpus is based on the value of JOB_GLIDEIN_Cpus once the routed job is matched to an
         # HTCondor slot
         SET GlideinCpusIsGood      int(MATCH_EXP_JOB_GLIDEIN_Cpus ?: "0") isnt error
         # Also used by JobGpus and JobMemory
-        SET JobIsRunning           (JobStatus =!= 1) && (JobStatus =!= 5) && GlideinCpusIsGood
+        SET JobIsRunning           $(MY.JobIsRunning) && GlideinCpusIsGood
+        COPY RequestCpus           OriginalCpus
         SET JobCpus                JobIsRunning ? int(MATCH_EXP_JOB_GLIDEIN_Cpus) : OriginalCpus
         SET RequestCpus            TARGET.TotalCpus ?: JobCpus
     endif
 
-    EVALMACRO test_xcount $(MY.xcount:0) > 1
-    if $(test_xcount)
-        cpus = $(MY.xcount)
-    else
-        EVALMACRO test_orig_RequestCpus $(MY.orig_RequestCpus:0) > 1
-        if $(test_orig_RequestCpus)
-            cpus = $(MY.orig_RequestCpus)
-         else
-            cpus = $(default_xcount:1)
-         endif
-    endif
-
-    EVALSET OriginalCpus           $(cpus)
-    EVALSET remote_SMPGranularity  $(cpus)
-    EVALSET remote_NodeNumber      $(cpus)
 
-    DEFAULT JobIsRunning           (JobStatus =!= 1) && (JobStatus =!= 5)
-    DEFAULT JobCpus                OriginalCpus
-    DEFAULT RequestCpus            OriginalCpus
 @jrt
 
 
 JOB_ROUTER_TRANSFORM_Gpus @=jrt
     # Request GPUs for whole node jobs (HTCONDOR-103)
     # If a whole node job requests GPUs and is matched to a machine with GPUs then set the job's RequestGPUs to all the
     # GPUs on that machine
+    REQUIREMENTS RequestGpus isnt undefined
```

I think we need something like the following:

```suggestion
    REQUIREMENTS RequestGpus isnt undefined || WantWholeNodes isnt undefined
```

Otherwise `WantWholeNode` requests won't make use of any of the host's GPUs

---

change is needed to make the code do what the comment says it does, your alteration breaks that

---

if we give GPUs to jobs that don't want them, it won't be possible to tell from condor_status that the GPUs are going unused

---

Ah I had thought the comment was wrong but I misread the original logic in `JOB_ROUTER_DEFAULTS` so nevermind on this one

---

> if we give GPUs to jobs that don't want them, it won't be possible to tell from condor_status that the GPUs are going unused

OTOH, giving an entire node with a GPU to a pilot that doesn't make use of the GPU is a waste. Though I would guess that such a case is protected against by most `START` expressions?